File: src/java/burp/em/ConcreteExpressionMap.java
package burp.em;

import java.util.ArrayList;
import java.util.List;

import burp.exp.RDFNodeConstant;
import burp.exp.Reference;
import burp.exp.Template;
import burp.iteration.Iteration;

public class ConcreteExpressionMap extends ExpressionMap {

	public List<String> generateValues(Iteration i) {
		List<String> l = new ArrayList<String>();
		
		if(expression instanceof RDFNodeConstant) {
			// It is assumed to be a string, otherwise the shapes
			// Would have caught the error.
			l.add(((RDFNodeConstant) expression).constant.toString());
			return l;
		}
		
		if(expression instanceof Template) {
			l.addAll(((Template) expression).values(i));
			return l;
		}
		
		if(expression instanceof Reference) {
			List<Object> values = ((Reference) expression).values(i);
			for(Object o : values)
				l.add(o.toString());
			return l;
		}
		
		throw new RuntimeException("Error generating language string.");
	}

}

File: src/main/java/burp/model/ConcreteExpressionMap.kt
package burp.model

import burp.reporting.BurpException
import burp.reporting.RmlError
import burp.vocabularies.RER

/**
 * 
 * A ConcreteExpressionMap is a concrete implementation of the abstract class
 * ExpressionMap for use in join conditions and logical views
 * 
 */
class ConcreteExpressionMap : ExpressionMap() {
    fun generateValues(i: Iteration, baseIRI: String): List<Any?> =
        when (val expr = expression) {
            is RDFNodeConstant -> mutableListOf<Any?>(expr.constant.toString())
            is Template -> ArrayList<Any?>(expr.values(i))
            is Reference -> expr.values(i)
            is FunctionExecution -> expr.values(i, baseIRI)
            else -> throw BurpException(
                RmlError(
                    "Unsupported type of values expression in Expression Map.",
                    expressionOrigin,
                    RER.UnsupportedMapping
                )
            )
        }
}